sklearn/metrics/_pairwise_distances_reduction.pyx

```diff
@@ -968,6 +968,88 @@ cdef class PairwiseDistancesArgKmin(PairwiseDistancesReduction):
         return np.asarray(self.argkmin_indices)
 
 
+cdef class GEMMComponent:
+    """Component for FastEuclidean alternatives wrapping the logic for the call to GEMM."""
+
+    cdef:
+        const DTYPE_t[:, ::1] X
+        const DTYPE_t[:, ::1] Y
+
+        ITYPE_t effective_n_threads
+        ITYPE_t chunks_n_threads
+        ITYPE_t dist_middle_terms_chunks_size
+
+        # Buffers for the `-2 * X_c @ Y_c.T` term computed via GEMM
+        vector[vector[DTYPE_t]] dist_middle_terms_chunks
+
+    def __init__(self,
+        DTYPE_t[:, ::1] X,
+        DTYPE_t[:, ::1] Y,
+        ITYPE_t effective_n_threads,
+        ITYPE_t chunks_n_threads,
+        ITYPE_t dist_middle_terms_chunks_size,
+    ):
+        self.X = X
+        self.Y = Y
+        self.effective_n_threads = effective_n_threads
+        self.chunks_n_threads = chunks_n_threads
+        self.dist_middle_terms_chunks_size = dist_middle_terms_chunks_size
+
+        self.dist_middle_terms_chunks = vector[vector[DTYPE_t]](self.effective_n_threads)
+
+    cdef void _parallel_on_X_parallel_init(self, ITYPE_t thread_num) nogil:
+        self.dist_middle_terms_chunks[thread_num].resize(self.dist_middle_terms_chunks_size)
+
+    cdef void _parallel_on_Y_init(self) nogil:
+        for thread_num in range(self.chunks_n_threads):
+            self.dist_middle_terms_chunks[thread_num].resize(
+                self.dist_middle_terms_chunks_size
+            )
+
+    cdef DTYPE_t * _compute_distances_on_chunks(
+        self,
+        ITYPE_t X_start,
+        ITYPE_t X_end,
+        ITYPE_t Y_start,
+        ITYPE_t Y_end,
```

Comment:
Will `GEMMComponent` need `self.X` or `self.Y` in the future? If not, we can pass all the data in here and not need `self.X` or `self.Y`:

```suggestion
        const DTYPE_t[:, ::1] X,
        ITYPE_t X_start,
        ITYPE_t X_end,
        const DTYPE_t[:, ::1] Y,
        ITYPE_t Y_start,
        ITYPE_t Y_end,
```

(I think having less state in `GEMMComponent` is better. )

Reply:
Well, actually I am wrong: they are also use to compute the squared norm vectors.

Moreover, we might want to pass other memory-views / buffers to define `X_c` and `Y_c` in some implementation when up-casting.

Hence, I am in favour of your proposal.

Reply:
(I wish multiple inheritance and/or traits were present in Cython.)

Reply:
Well, thinking of it again.

Probably it is better to keep `GEMMComponent.{X,Y}` in to encapsulate [the subsequent upcasting logic](https://github.com/jjerphan/scikit-learn/blob/distance-metrics-32bit/sklearn/metrics/_pairwise_distances_reduction.pyx.tp#L1205-L1213) in this component as well.

WDYT?

Reply:
Seeing the upcast logic, I thinks better to keep the `GEMMComponent.{X,Y}`.